File: android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSRemoteVideoTrackAction.kt
package live.hms.hmssdk_flutter.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.hmssdk_flutter.HMSExceptionExtension
import live.hms.hmssdk_flutter.HMSSimulcastLayerDefinitionExtension
import live.hms.hmssdk_flutter.HMSSimulcastLayerExtension
import live.hms.video.media.tracks.HMSRemoteVideoTrack
import live.hms.video.sdk.HMSSDK
import live.hms.video.sdk.models.HMSRoom
import live.hms.video.utils.HmsUtilities

class HMSRemoteVideoTrackAction {

    companion object {
        fun remoteVideoTrackActions(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK) {
            when (call.method) {
                "set_simulcast_layer" -> setSimulcastLayer(call, result, hmssdk)
                "get_layer" -> getLayer(call, result, hmssdk)
                "get_layer_definition" -> getLayerDefinition(call, result, hmssdk)
            }
        }

        private fun setSimulcastLayer(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK) {
            val trackId = call.argument<String>("track_id") ?:
            run {
                HMSErrorLogger.logError("setSimulcastLayer", "trackId is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("trackId is null in setSimulcastLayer"))
                return
            }
            val simulcastLayer = call.argument<String>("layer") ?:
            run {
                HMSErrorLogger.logError("setSimulcastLayer", "simulcastLayer is null", "Parameter Error")
                result.success(HMSExceptionExtension.getError("simulcastLayer is null in setSimulcastLayer"))
                return
            }

            val room = hmssdk.getRoom()
            room  ?:
            run{
                HMSErrorLogger.logError("setSimulcastLayer", "room is null", "Null Error")
                result.success(HMSExceptionExtension.getError("room is null in setSimulcastLayer"))
                return
            }
            val track: HMSRemoteVideoTrack? = HmsUtilities.getVideoTrack(trackId, room) as HMSRemoteVideoTrack?
            track ?:
            run{
                HMSErrorLogger.logError("setSimulcastLayer", "Can't find track with trackId:$trackId", "Null Error")
                result.success(HMSExceptionExtension.getError("Can't find track with trackId:$trackId in setSimulcastLayer"))
                return
            }
            track.setLayer(HMSSimulcastLayerExtension.getLayerFromString(layer = simulcastLayer))
            result.success(null)
        }

        private fun getLayer(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK) {
            val trackId = call.argument<String>("track_id") ?:
            run {
                HMSErrorLogger.logError("getLayer", "trackId is null", "Parameter Error")
                result.success(null)
                return
            }
            val room: HMSRoom = hmssdk.getRoom()
                ?: run {
                    HMSErrorLogger.logError("setSimulcastLayer", "room is null", "Null Error")
                    result.success(null)
                    return
                }
            val track: HMSRemoteVideoTrack? = HmsUtilities.getVideoTrack(trackId, room) as HMSRemoteVideoTrack?
            if (track != null) {
                result.success(HMSSimulcastLayerExtension.getStringFromLayer(track.getLayer()))
            }
            result.success(null)
        }

        private fun getLayerDefinition(call: MethodCall, result: MethodChannel.Result, hmssdk: HMSSDK) {
            val trackId = call.argument<String>("track_id") ?:
            run {
                HMSErrorLogger.logError("getLayerDefinition", "trackId is null", "Parameter Error")
                result.success(null)
                return
            }

            val room: HMSRoom = hmssdk.getRoom()
           ?: run{
               HMSErrorLogger.logError("setSimulcastLayer", "room is null", "Null Error")
               result.success(null)
                return
            }
            val track: HMSRemoteVideoTrack? = HmsUtilities.getVideoTrack(trackId, room) as HMSRemoteVideoTrack?
            val hashMap = ArrayList<HashMap<String, Any?>>()
            if (track != null) {
                track.getLayerDefinition().forEach {
                    hashMap.add(HMSSimulcastLayerDefinitionExtension.toDictionary(it))
                }
                result.success(hashMap)
            }
            result.success(null)
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Actions/HMSRemoteVideoTrackAction.swift
//
//  HMSRemoteVideoTrackAction.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 19/12/22.
//

import Foundation
import HMSSDK

class HMSRemoteVideoTrackExtension {

    static func remoteVideoTrackActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK) {

        switch call.method {
        case "set_simulcast_layer":
            setSimulcastLayer(call, result, hmsSDK)
        case "get_layer":
            getLayer(call, result, hmsSDK)
        case "get_layer_definition":
            getLayerDefinition(call, result, hmsSDK)
        default:
            result(FlutterMethodNotImplemented)
        }

    }

    static private func setSimulcastLayer(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK) {
        let arguments = call.arguments as![AnyHashable: Any]

        guard let trackIdString = arguments["track_id"] as? String,
              let trackId = HMSUtilities.getTrack(for: trackIdString, in: hmsSDK.room!),
              let layer = arguments["layer"] as? String
        else {
            result(HMSErrorExtension.getError("No track ID or HMSSimulcastLayer found in \(#function)"))
            return
        }

        if let track = trackId as? HMSRemoteVideoTrack {
            track.layer = HMSSimulcastLayerDefinitionExtension.getLayerFromString(layer: layer)
            result(nil)
        } else {
            result(HMSErrorExtension.getError("Track Id must be HMSRemoteVideoTrack Type \(#function)"))
        }
    }

    static private func getLayer(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK) {
        let arguments = call.arguments as![AnyHashable: Any]

        guard let trackIdString = arguments["track_id"] as? String,
              let trackId = HMSUtilities.getTrack(for: trackIdString, in: hmsSDK.room!)
        else {
            result(nil)
            return
        }
        if let track = trackId as? HMSRemoteVideoTrack {
            result(HMSSimulcastLayerDefinitionExtension.getStringFromLayer(layer: track.layer))
        } else {
            result(nil)
        }
    }

    static private func getLayerDefinition(_ call: FlutterMethodCall, _ result: @escaping FlutterResult, _ hmsSDK: HMSSDK) {
        let arguments = call.arguments as![AnyHashable: Any]

        guard let trackIdString = arguments["track_id"] as? String,
              let trackId = HMSUtilities.getTrack(for: trackIdString, in: hmsSDK.room!)
        else {
            result(nil)
            return
        }
        var dict = [[String: Any]]()
        if let track = trackId as? HMSRemoteVideoTrack {
            track.layerDefinitions?.forEach { layerDefinition in
                dict.append(HMSSimulcastLayerDefinitionExtension.toDictionary(layerDefinition))
            }
            result(dict)
        } else {
            result(nil)
        }
    }

}
